packages/block-editor/src/components/block-list/index.js

```diff
@@ -41,13 +46,55 @@ import {
 	DEFAULT_BLOCK_EDIT_CONTEXT,
 } from '../block-edit/context';
 
-const elementContext = createContext();
+const componentsContext = createContext();
 
 export const IntersectionObserver = createContext();
 const pendingBlockVisibilityUpdatesPerRegistry = new WeakMap();
 
+function useRootPortal( component ) {
+	const components = useContext( componentsContext );
+
+	useLayoutEffect( () => {
+		if ( ! components ) return;
+		components.add( component );
+		return () => {
+			components.delete( component );
+		};
+	} );
+}
+
+function Components( { subscriber, components } ) {
+	const [ , forceRender ] = useReducer( () => ( {} ) );
+	subscriber.current = forceRender;
+	return createElement( Fragment, null, ...components.current );
+}
+
+function ComponentRenderer( { children } ) {
+	const subscriber = useRef( () => {} );
+	const components = useRef( new Set() );
+	return (
+		<componentsContext.Provider
+			value={ useMemo(
+				() => ( {
+					add( component ) {
+						components.current.add( component );
+						subscriber.current();
+					},
+					delete( component ) {
+						components.current.delete( component );
+						subscriber.current();
```

Comment:
Whenever a component that `useRootPortal` re-renders, it will delete the content and re-add it in `useLayoutEffect`, thus calling `subscriber` twice, which is just a `forceRender` function in the `Components` component. This results in a double-render for the portal component for every render of the consumer's component. While React is typically very fast, this does feel like an unnecessary overhead.

Reply:
I'll fix this.